internal/website/gatherexamples/gatherexamples.go

```diff
@@ -0,0 +1,271 @@
+		os.Exit(1)
+	}
+}
+
+const gatherLoadMode packages.LoadMode = packages.NeedName |
+	packages.NeedFiles |
+	packages.NeedTypes |
+	packages.NeedSyntax |
+	packages.NeedTypesInfo
+
+func gather(pkgs []*packages.Package) map[string]string {
+	examples := make(map[string]string)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			for _, decl := range file.Decls {
+				// Determine whether this declaration is an example function.
+				fn, ok := decl.(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+				if !ok || !strings.HasPrefix(fn.Name.Name, "Example") || len(fn.Type.Params.List) > 0 || len(fn.Type.Params.List) > 0 {
+					continue
+				}
+
+				// Gather sorted list of imported packages.
+				usedPackages := make(map[string]string)
+				ast.Inspect(fn.Body, func(node ast.Node) bool {
+					id, ok := node.(*ast.Ident)
+					if !ok {
+						return true
+					}
+					refPkg, ok := pkg.TypesInfo.ObjectOf(id).(*types.PkgName)
+					if !ok {
+						return true
+					}
+					overrideName := ""
+					if id.Name != refPkg.Imported().Name() {
+						overrideName = id.Name
+					}
+					usedPackages[refPkg.Imported().Path()] = overrideName
+					return true
+				})
+
+				// Format example into string.
+				sb := new(strings.Builder)
+				err := printer.Fprint(sb, pkg.Fset, &printer.CommentedNode{
+					Node:     fn.Body,
+					Comments: file.Comments,
+				})
+				if err != nil {
+					panic(err) // will only occur for bad invocations of Fprint
+				}
+				exampleCode := rewriteBlock(sb.String())
+				if len(usedPackages) > 0 {
+					exampleCode = formatImports(usedPackages) + "\n\n" + exampleCode
+				}
+
+				pkgPath := strings.TrimSuffix(pkg.PkgPath, "_test")
+				exampleName := pkgPath + "." + fn.Name.Name
+				examples[exampleName] = exampleCode
+			}
+		}
+	}
+	return examples
+}
+
+// rewriteBlock reformats a Go block statement for display as an example.
+func rewriteBlock(block string) string {
+	// Trim block.
+	block = strings.TrimPrefix(block, "{")
+	block = strings.TrimSuffix(block, "}")
+
+	// Rewrite line-by-line.
+	sb := new(strings.Builder)
+rewrite:
+	for len(block) > 0 {
+		var line string
+		line, block = nextLine(block)
+
+		// Dedent line.
+		// TODO(light): In the case of a multi-line raw string literal,
+		// this can produce incorrect rewrites.
+		line = strings.TrimPrefix(line, "\t")
+
+		// Check if this is a special line.
+		start := strings.IndexFunc(line, func(r rune) bool { return r != ' ' && r != '\t' })
+		if start == -1 {
+			// Blank.
+			sb.WriteString(line)
+			sb.WriteByte('\n')
+			continue
+		}
+		indent := line[:start]
+		switch line[start:] {
+		case "// Variables set up elsewhere:":
+			// Skip lines until we hit a blank line.
+			for len(block) > 0 {
+				var next string
+				next, block = nextLine(block)
+				if strings.TrimSpace(next) == "" {
+					break
+				}
+			}
+		case "// Ignore unused variables for example:":
+			// Ignore remaining lines.
+			break rewrite
```

Comment:
Is the break and label necessary, or can you just return `strings.TrimSpace(sb.String())` here?

Reply:
The loop can exit without hitting such a comment line, so both control flow paths need to do the `strings.TrimSpace`. This seemed clearer to me.